pkg/logging/http.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package logging
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"net/http"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	package_http "github.com/thanos-io/thanos/pkg/server/http"
+)
+
+type HTTPServerMiddleware struct {
+	opts   *options
+	logger log.Logger
+}
+
+func (c *HTTPServerMiddleware) PostCall(name string, start time.Time, wrapped *package_http.ResponseWriter, r *http.Request) {
+	reqID := r.Header.Get("X-Request-ID")
+
+	logger := log.With(c.logger, "http.request_logging_server", name)
+	logger = log.With(logger, "http.request-id", reqID)
+	status := wrapped.Status()
+	if status >= 200 && status < 500 {
+		logger = log.With(logger, "http.code", fmt.Sprintf("%d", status))
+	}
+	// server error, log them.
+	if status >= 500 && status < 600 {
+		logger = log.With(logger, "http.error", fmt.Sprintf("http.code: %v", status))
+	}
+	logger = log.With(logger, "http.time_ms", fmt.Sprintf("%v", durationToMilliseconds(time.Since(start))))
+	logger = c.opts.levelFunc(logger, status)
+	logger = log.With(logger, "protocol", "http")
```

Comment:
Use the `protocol` at higher level. Maybe add the `http` in the constructor, wrap the logger in the constructor itself.